File: example-kmp-app/androidApp/src/main/kotlin/com/kmpfoo/android/ui/tabhost/RootTabHostView.kt
package com.kmpfoo.android.ui.tabhost

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import co.early.n8.NavigationModel
import co.early.n8.NavigationState
import co.early.n8.TabHostLocation
import com.kmpfoo.android.OG
import com.kmpfoo.ui.navigation.Location
import com.kmpfoo.ui.navigation.TabHostId

private val navModel by lazy {
    OG[NavigationModel::class] as NavigationModel<Location, TabHostId>
}

@Composable
fun RootTabHostView(
    navigationState: NavigationState<Location, TabHostId>,
    wrappingTabHosts: List<TabHostLocation<TabHostId>>,
    depth: Int,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit
) {

    val tabHostLocation = wrappingTabHosts[depth]
    val tabHost = navigationState.locateTabHost(tabHostLocation.tabHostId)

    requireNotNull(tabHost){
        "this should never be null, it means the tabHost does not exist in the navigation graph yet, check your code"
    }

    val nestedContent = if (wrappingTabHosts.size > depth + 1) {
        { RootTabHostView(navigationState, wrappingTabHosts, depth + 1, modifier, content) }
    } else {
        content
    }

    when (tabHostLocation.tabHostId){
        TabHostId.EuropeTabHost -> {
            TabHostEurope(tabHost.tabHistory.last(), navModel, modifier, nestedContent)
        }
        TabHostId.GlobalTabHost -> {
            TabHostGlobal(tabHost.tabHistory.last(), navModel, modifier, nestedContent)
        }
    }
}

File: example-kmp-app/iosApp/iosApp/ui/tabhost/RootTabHostView.swift
import SwiftUI
import shared

struct RootTabHostView: View {
    
    let n8: NavigationModel<Location, TabHostId>
    let wrappingTabHosts: [TabHostLocation<TabHostId>]
    let tabHostLocation: TabHostLocation<TabHostId>
    let tabHost: NavigationTabHost<Location, TabHostId>
    let depth: Int
    let content: () -> AnyView
    
    init(
        n8: NavigationModel<Location, TabHostId>,
        depth: Int = 0,
        content: @escaping () -> AnyView
    ) {
        self.n8 = n8
        self.wrappingTabHosts = n8.state.hostedBy
        self.tabHostLocation = wrappingTabHosts[depth]
        self.tabHost = n8.state.locateTabHost(tabHostId:tabHostLocation.tabHostId!)!
        self.depth = depth
        self.content = content
    }
    
    var body: some View {

        let isLast = wrappingTabHosts.count <= depth + 1
        let nestedContent: () -> AnyView = isLast ? content : {
            AnyView(
                RootTabHostView(
                    n8: n8,
                    depth: depth + 1,
                    content: content
                )
            )
        }
        
        switch tabHostLocation.tabHostId {
            case is TabHostId.EuropeTabHost:
                TabHostEurope(selectedTabIndex: Int(truncating:tabHost.tabHistory.last!), content: nestedContent)
            case is TabHostId.GlobalTabHost:
                TabHostGlobal(selectedTabIndex: Int(truncating:tabHost.tabHistory.last!), content: nestedContent)
            default: Text("Unknown tabhost \(tabHostLocation.tabHostId!)")
        }
    }
}
